{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Designing on-target and mismatched sgRNAs for tirated CRISPRi\n",
    "\n",
    "This code finds the basepairing region for single guide RNAs for CRISPR interference and permutes the on-target sgRNA to the mismatch patterns discussed in Mathis et al. 2020. Outputs the sgRNA sequence incorporated into a primer to clone the basepairing region into the plasmid background used in this study.\n",
    "\n",
    "Central Inputs:\n",
    "An array of E. coli gene names, which cann be modified for other organisms, but you need the whole organism fasta file which can be limiting.\n",
    "Your target PAM sequence. Should not change unless you are using a different Cas9. The current Cas9 is S. pyogenes, which is in the pdCas9-bacteria plamid. Addgene-Plasmid:44249\n",
    "Primer_H is the homology region on the target insert vector for the BP region of the gRNA. These BP regions can be inserted using inverse PCR. See Hawkins et al. 2015 for details on molecular biology steps.\n",
    "\n",
    "Details: \n",
    "Searches e_coli_gene_sequences_NC_000913p3.txt for the genes of interest and extracts the fasta sequence (these are all nontemplate sequences.\n",
    "For each fasta sequence, all PAM sequences are found. The PAM closest to the n-terminus is selected.\n",
    "Takes reverse complement of the 20BPs downstream of the PAM sequence. This is the sgRNA sequence targeting the non-template strand.\n",
    "Checks sgRNA and sgRNA reverse complement for BsaI sites. If BsaI sites are present, it will be shown in output.\n",
    "BLASTs BP target gene sequence. Homology with other genes suggests sgRNA need to be redisgned.\n",
    "\n",
    "Outputs:\n",
    "CRISPRi_Design_*_.txt which contains all the relevant information about the sgRNAs.\n",
    "fasta_order containing full sequences needed for sgRNA library synthesis\n",
    "\n",
    "Written by Andrew Mathis\n",
    "\n",
    "Updated 23 July 2020"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "from Bio.Seq import Seq\n",
    "from Bio import Restriction\n",
    "import numpy as np\n",
    "from collections import Counter\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_path = '../inputs/20200207_selected_genes_chem.csv' # Path to selected gene information\n",
    "genome_path = '../inputs/e_coli_genome_CP032667p1.txt' # Path to genome txt file for the E. coli genome\n",
    "genome_fasta_path = '../inputs/e_coli_gene_sequences_NC_000913p3.txt' # Path to fasta file for the target genome\n",
    "sgRNA_path = 'Gradients/' # General path for sgRNA sequences\n",
    "output_path = 'Outputs/' # General path for output files\n",
    "blast_path = 'Blast/' # General path to BLAST output files\n",
    "\n",
    "PAM_NT = 'C|C|N' # Formatted PAM targeting sequence\n",
    "\n",
    "primer_H = 'GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC' # Homology region to target plasmid\n",
    "Promoter = 'TTGACAGCTAGCTCAGTCCTAGGTATAATACTAGT' # sgRNA promoter sequence\n",
    "# Full sgRNA flanking sequnece\n",
    "LoopTerm = 'GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTTGAAG'\n",
    "\n",
    "mut_pos1 = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 15, 18, 19] # Index for single mismatches in sgRNAs\n",
    "mut_pairs = [[3, 5], [1, 7], [6, 8], [2, 9], [8, 18]] # Indices for double mismatches in sgRNAs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Central functions for analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Finds 3 potential sgRNAs per gene of interest. Two sgRNAs as close as possible to the 3' prime (1 to 150 BP) end\n",
    "# without overlap and one at least 200 BP from 3' end. Exclused sgRNAs containing four or more consecutive thymidines\n",
    "# or a GC content < 45% or > 80%. Cutoffs come from CRISPR-ERA design and Gilbert et al. 2014.\n",
    "# Input -- Gene_file: Gene sequence information from E. coli genome\n",
    "#          Gene_list: List of genes to target with CRISPRi\n",
    "# Output -- df_parent: Potential sgRNA sequences formatted for mutagenesis\n",
    "#           sgRNA_parent_blast: Potential sgRNA sequences formatted for BLAST\n",
    "def pick_sg(Gene_file, Gene_list):\n",
    "    sgRNA_parent = []\n",
    "    sgRNA_parent_blast = []\n",
    "    parent_db = {}\n",
    "    for file_name in Gene_file:\n",
    "        Gene = file_name.split('_')[-2]\n",
    "        if Gene in Gene_list:\n",
    "            with open(file_name) as FI:\n",
    "                PAM = 0\n",
    "                sg1_count = 0 # Number of sgRNAs close to start of gene\n",
    "                sg2_count = 0 # Number of sgRNAs far from start of gene\n",
    "                for i, line in enumerate(FI):\n",
    "                    if sg1_count >= 2 and sg2_count >=1: # Search complete\n",
    "                        break\n",
    "                    if i > 0:\n",
    "                        sp = line.split('\\t')\n",
    "                        sg = sp[1]\n",
    "                        PAM_index = int(sp[3])\n",
    "                        PAM_seq = sp[2]\n",
    "                        if 'TTTT' not in sg: # This sequence is unfavorable for CRISPRi\n",
    "                            counts = Counter(sg) # Counts frequency of each letter\n",
    "                            CG_ratio = np.true_divide((counts['C'] + counts['G']), sum(counts.values()))\n",
    "                            if PAM_index > PAM + 20 or PAM == 0:\n",
    "                                if CG_ratio >= 0.45 and CG_ratio <= 0.80: # Optimal nucleotide ratio range\n",
    "                                    if sg1_count < 2 and PAM_index <= 150: # First 2 sgRNAs must be near start\n",
    "                                        Gene_ID = Gene + '_' + str(sg1_count+1) + '_' + str(PAM_index)\n",
    "                                        PAM = PAM_index\n",
    "                                        sg1_count += 1\n",
    "                                        # List for mutagenesis\n",
    "                                        parent_db[Gene_ID] = {}\n",
    "                                        sgRNA_parent.append([Gene_ID, sg])\n",
    "\n",
    "                                        # List for BLAST\n",
    "                                        sgRNA_parent_blast.append('> '+ Gene_ID)\n",
    "                                        sgRNA_parent_blast.append(sg)\n",
    "\n",
    "                                        # Database\n",
    "                                        parent_db[Gene_ID]['Sequence'] = sg\n",
    "                                        parent_db[Gene_ID]['PAM_seq'] = PAM_seq\n",
    "                                        parent_db[Gene_ID]['PAM_index'] = PAM_index\n",
    "\n",
    "                                    elif sg2_count < 1 and PAM_index >= 200: # Third sgRNA >=200BP from start\n",
    "                                        Gene_ID = Gene + '_3_' + str(PAM_index)\n",
    "                                        sg2_count += 1\n",
    "                                        sgRNA_parent.append([Gene_ID, sg])\n",
    "                                        sgRNA_parent_blast.append('> '+ Gene_ID)\n",
    "                                        sgRNA_parent_blast.append(sg)\n",
    "                                        parent_db[Gene_ID] = {}\n",
    "\n",
    "                                        # Database\n",
    "                                        parent_db[Gene_ID]['Sequence'] = sg\n",
    "                                        parent_db[Gene_ID]['PAM_seq'] = PAM_seq\n",
    "                                        parent_db[Gene_ID]['PAM_index'] = PAM_index\n",
    "    df_parent = pd.DataFrame.from_dict(parent_db).T\n",
    "    \n",
    "    return df_parent, sgRNA_parent_blast\n",
    "\n",
    "\n",
    "# Finds 3 potential sgRNAs per gene of interest, specificially for genes where one or more sgRNAs from their initial\n",
    "# list failed at the BLAST step in analysis. Two sgRNAs as close as possible to the 3' prime (1 to 150 BP) end\n",
    "# without overlap and one at least 200 BP from 3' end. Exclused sgRNAs containing four or more consecutive thymidines\n",
    "# or a GC content < 45% or > 80%. Cutoffs come from CRISPR-ERA design and Gilbert et al. 2014.\n",
    "# Input -- Gene_file: Gene sequence information from E. coli genome\n",
    "#          Fail_List: List of genes to target with CRISPRi along with information on previously failed PAM locations\n",
    "# Output -- df_parent: Potential sgRNA sequences formatted for mutagenesis\n",
    "#           sgRNA_parent_blast: Potential sgRNA sequences formatted for BLAST\n",
    "def pick_sg_it(Gene_file, Fail_list):\n",
    "    sgRNA_parent = []\n",
    "    sgRNA_parent_blast = []\n",
    "    parent_db = {}\n",
    "    Gene_list = []\n",
    "    PAM_bad = []\n",
    "    for gene in Fail_list:\n",
    "        sp = gene.split('_')\n",
    "        Gene_list.append(sp[0])\n",
    "        PAM_bad.append(sp[-1])\n",
    "    for file_name in Gene_file:\n",
    "        Gene = file_name.split('_')[-2]\n",
    "        if Gene in Gene_list:\n",
    "            PAM_B = PAM_bad[Gene_list.index(Gene)]\n",
    "            with open(file_name) as FI:\n",
    "                PAM = 0\n",
    "                sg1_count = 0 # Number of sgRNAs close to start of gene\n",
    "                sg2_count = 0 # Number of sgRNAs far from start of gene\n",
    "                for i, line in enumerate(FI):\n",
    "                    if sg1_count >= 2 and sg2_count >=1: # Search complete\n",
    "                        break\n",
    "                    if i > 0:\n",
    "                        sp = line.split('\\t')\n",
    "                        sg = sp[1]\n",
    "                        PAM_index = int(sp[3])\n",
    "                        PAM_seq = sp[2]\n",
    "                        if 'TTTT' not in sg: # This sequence is unfavorable for CRISPRi\n",
    "                            counts=Counter(sg) # Counts frequency of each letter\n",
    "                            CG_ratio = np.true_divide((counts['C'] + counts['G']), sum(counts.values()))\n",
    "                            if CG_ratio >= 0.45 and CG_ratio <= 0.80: # Optimal nucleotide ratio range\n",
    "                                if PAM == 0 or PAM_index > PAM + 20:\n",
    "                                    if PAM_index != int(PAM_B):\n",
    "                                        if sg1_count < 2 and PAM_index <= 150: # First 2 sgRNAs must be near start\n",
    "                                            Gene_ID = Gene +'_'+str(sg1_count+1)+'_'+str(PAM_index)\n",
    "                                            PAM = PAM_index\n",
    "                                            sg1_count += 1\n",
    "                                            # List for mutagenesis\n",
    "                                            parent_db[Gene_ID] = {}\n",
    "                                            sgRNA_parent.append([Gene_ID, sg])\n",
    "\n",
    "                                            # List for BLAST\n",
    "                                            sgRNA_parent_blast.append('> '+ Gene_ID)\n",
    "                                            sgRNA_parent_blast.append(sg)\n",
    "\n",
    "                                            # Database\n",
    "                                            parent_db[Gene_ID]['Sequence'] = sg\n",
    "                                            parent_db[Gene_ID]['PAM_seq'] = PAM_seq\n",
    "                                            parent_db[Gene_ID]['PAM_index'] = PAM_index\n",
    "\n",
    "                                        elif sg2_count < 1 and PAM_index >= 200: # Third sgRNA >=200BP from start\n",
    "                                            Gene_ID = Gene+'_3_'+str(PAM_index)\n",
    "                                            sg2_count += 1\n",
    "                                            sgRNA_parent.append([Gene_ID, sg])\n",
    "                                            sgRNA_parent_blast.append('> '+ Gene_ID)\n",
    "                                            sgRNA_parent_blast.append(sg)\n",
    "                                            parent_db[Gene_ID] = {}\n",
    "\n",
    "                                            # Database\n",
    "                                            parent_db[Gene_ID]['Sequence'] = sg\n",
    "                                            parent_db[Gene_ID]['PAM_seq'] = PAM_seq\n",
    "                                            parent_db[Gene_ID]['PAM_index'] = PAM_index\n",
    "    df_parent = pd.DataFrame.from_dict(parent_db).T\n",
    "    \n",
    "    return df_parent, sgRNA_parent_blast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Determines if the sgRNA homology region is unique enough to use in CRISPRi\n",
    "# Input -- gene: Gene of interest for CRISPRi\n",
    "#          seq: BLASTed alignment sequence\n",
    "#          PAM: Potential PAM associated with the BLASTed alignment sequence\n",
    "#          parent: Parent sgRNA sequence\n",
    "#          MM: Mismatch number BP homology between the parent and BLASTed alignment\n",
    "#          df_parent: Potential sgRNA sequences formatted for mutagenesis\n",
    "# Output -- Pass/Fail: A pass indicates the sgRNA in question will likely lack off target effects. A fail excludes\n",
    "#                      this sgRNA from future analysis\n",
    "def blast_test(gene, seq, PAM, parent, MM, df_parent):\n",
    "    # Upstream analysis already checks fully on-target sgRNAs\n",
    "    if MM == 20:\n",
    "        # Ensure to make sure the algorithm is extracting the correct sequences\n",
    "        if df_parent['Sequence'].loc[gene] == seq and df_parent['PAM_seq'].loc[gene] == PAM:\n",
    "            return 'pass'\n",
    "        else:\n",
    "            print('Algorithm grabbed wrong sequence or PAM for 100% 20N mismatch for: ' + gene)\n",
    "            return 'fail'\n",
    "    # Is there an adjacent PAM?\n",
    "    if PAM[0] == 'C' and PAM[1] == 'C':\n",
    "        # Does the guide have at least two mismatches? If not, off-target potential is high\n",
    "        if MM >= 18:\n",
    "            return 'fail'\n",
    "        # Is at least one mismatch in region 1 (critical region for binding)? If not, off-target potential is high\n",
    "        else:\n",
    "            if seq[-8:] == parent[-8:]:\n",
    "                return 'fail'\n",
    "            else:\n",
    "                return 'pass'\n",
    "    else:\n",
    "        return 'pass'\n",
    "\n",
    "    \n",
    "# Determine which sgRNAs with high sequence identity to other regions of the E. coli genome have high potential for\n",
    "# off-target effects\n",
    "# Input -- Blast_count: Number of BLAST hits per potential sgRNA\n",
    "#          df_15: BLAST hits with at least 75% similarity to a given sgRNA\n",
    "#          df_parent: Potential sgRNA sequences formatted for mutagenesis\n",
    "# Output -- Gene_pass: Genes without any potential sgRNAs with a high chance of off-target effects\n",
    "#           Gene_fail: Genes with at least one potential sgRNA with a high chance of off-target effects\n",
    "def blast_check(Blast_count, df_15, df_parent):\n",
    "    Gene_fail = []\n",
    "    Gene_pass_temp = []\n",
    "    for gene in Blast_count.keys():\n",
    "        if gene.split('_')[0] in df_genes.index:\n",
    "            if Blast_count[gene] > 1:\n",
    "                starts = df_15['LocF'].loc[gene]\n",
    "                stops = df_15['LocR'].loc[gene]\n",
    "                MM = df_15['BP_Match'].loc[gene]\n",
    "                if Counter(MM)[20] == 1:\n",
    "                    for i, start in enumerate(starts):\n",
    "                        stop = stops[i]\n",
    "                        if start < stop:\n",
    "                            bSeq = EC_Genome[start-1:stop] # Extract guide sequence\n",
    "                            PAM = Seq(EC_Genome[stop:stop+3]).reverse_complement()\n",
    "                            parent = df_parent['Sequence'].loc[gene]\n",
    "                            test = blast_test(gene, bSeq, PAM, parent, MM[i], df_parent) # Test for off-target effects\n",
    "                        else:\n",
    "                            PAM = EC_Genome[stop-4:stop-1]\n",
    "                            bSeq = Seq(EC_Genome[stop-1:start])\n",
    "                            bSeq = bSeq.reverse_complement()\n",
    "                            parent = df_parent['Sequence'].loc[gene]\n",
    "                            test = blast_test(gene, bSeq, PAM, parent, MM[i], df_parent) # Test for off-target effects\n",
    "\n",
    "                        if test == 'fail':\n",
    "                            Gene_fail.append(gene)\n",
    "                        else:\n",
    "                            Gene_pass_temp.append(gene)\n",
    "                else:\n",
    "                    Gene_fail.append(gene)\n",
    "    Gene_fail = list(set(Gene_fail))\n",
    "    Gene_pass_temp = list(set(Gene_pass_temp))\n",
    "    Gene_pass = []\n",
    "    for g in Gene_pass_temp:\n",
    "        if g not in Gene_fail:\n",
    "            Gene_pass.append(g)\n",
    "            \n",
    "    return Gene_fail, Gene_pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adds mismatches to on-target sgRNAs to create titrated CRISPRi sgRNA library and formats sgRNAs for iPCR\n",
    "# Input -- i: Index of the sgRNA of interest in final list\n",
    "#          sgRNA: List of all sgRNAs\n",
    "#          sgRNA_name: ID for the sgRNA of interest\n",
    "#          srt_I: Index of first mismatch to add to compounding mismatch guides\n",
    "#          stp_I: Index of last mismatch to add to compounding mismatch guides\n",
    "#          Promoter: Front-end flanking sequence for sgRNA\n",
    "#          LoopTerm: Back-end flanking sequence for sgRNA\n",
    "#          mut_pos1: Index for single mismatch additions\n",
    "#          mut_paris: Indices for double mismatch additions\n",
    "#          output_path: General path for output files\n",
    "# Output -- permute: Full list of permutes sgRNAs for titrated CRISPRi\n",
    "def permute_sgRNA(i, sgRNA, sgRNA_name, srt_I, stp_I, Promoter, LoopTerm, mut_pos1, mut_pairs, output_path):\n",
    "    Sub = {}\n",
    "    permute = {}\n",
    "    sgRNA_list = list(sgRNA) # Makes each nucleotide an index in the list\n",
    "    sgRNA_list_B = copy.deepcopy(sgRNA_list)\n",
    "    # Homologous mismatches to mimize emergent off-target effects\n",
    "    Sub['T'] = 'A'\n",
    "    Sub['A'] = 'T'\n",
    "    Sub['C'] = 'G'\n",
    "    Sub['G'] = 'C'\n",
    "    Gene_loc = sgRNA_name.split('_')[1]\n",
    "   \n",
    "    if Gene_loc != '2':\n",
    "        oligo1 = Promoter + sgRNA + LoopTerm\n",
    "        permute['C'] = oligo1\n",
    "        \n",
    "        for i, BP in enumerate(sgRNA_list):\n",
    "            if i in mut_pos1:\n",
    "                sgRNA_Wiess = list(sgRNA)\n",
    "                sgRNA_Wiess[i] = Sub[BP]\n",
    "                HR_Weiss = ''.join(sgRNA_Wiess)\n",
    "                oligo1 = Promoter + HR_Weiss + LoopTerm\n",
    "                permute['W_MM'+str(i+1)] = oligo1\n",
    "                \n",
    "            if i == 0:\n",
    "                sgRNA_list_B = sgRNA_list\n",
    "                sgRNA_list_B[i] = Sub[BP]\n",
    "            if i >= srt_I and i <= stp_I:\n",
    "                sgRNA_list_B[i] = Sub[BP]\n",
    "                HR_bikard = ''.join(sgRNA_list_B)\n",
    "                oligo2 = Promoter + HR_bikard + LoopTerm\n",
    "                permute['B_MM'+str(i+1)] = oligo2\n",
    "                \n",
    "        for pair in mut_pairs:\n",
    "            j, k = pair\n",
    "            sgRNA_dub = list(sgRNA)\n",
    "            sgRNA_dub[j] = Sub[sgRNA_dub[j]]\n",
    "            sgRNA_dub[k] = Sub[sgRNA_dub[k]]\n",
    "            HR_double = ''.join(sgRNA_dub)\n",
    "            oligo3 = Promoter + HR_double + LoopTerm\n",
    "            permute['D_MM' + str(j+1) + '-' + str(k+1)] = oligo3\n",
    "            \n",
    "    return permute"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import gene list\n",
    "df_genes = pd.read_csv(gene_path, index_col=0, parse_dates=True)\n",
    "Gene_list = list(df_genes.index)\n",
    "\n",
    "# Import E. coli genome\n",
    "EC_Genome = ''\n",
    "with open(genome_path, 'r') as FI:\n",
    "    for i, line in enumerate(FI):\n",
    "        if line[0] != '>':\n",
    "            EC_Genome = EC_Genome + line.rstrip('\\n')\n",
    "\n",
    "# Open the whole genome fasta file and load into memory, which likely computationally limits the size of genome\n",
    "MG1655_Genome = []\n",
    "with open(genome_fasta_path, 'r') as FI:\n",
    "    for line in FI:\n",
    "        MG1655_Genome.append(line.rstrip('\\n'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Find all possible sgRNAs per gene on non-template strand\n",
    "\n",
    "Searches the sequence of all genes of interest for potential sgRNA binding sites adjacent to a PAM motif. Collects and outputs all possible sgRNA locations and sequences for each gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_order = []\n",
    "gene_sequence = []\n",
    "# Extract the gene sequences associate with the genes of interest\n",
    "for i, line in enumerate(MG1655_Genome):\n",
    "    for gene in Gene_list:\n",
    "        gene_seq_temp = []\n",
    "        k = 1\n",
    "        if line.find('gene=' + gene) != -1: # Identify gene of interest\n",
    "            gene_order.append(gene)\n",
    "            while MG1655_Genome[i + k][0] != '>': # For the entire length of one gene\n",
    "                gene_seq_temp.append(MG1655_Genome[i + k])\n",
    "                k = k + 1\n",
    "            gene_seq_temp_flat = ''.join(gene_seq_temp)\n",
    "            gene_sequence.append(gene_seq_temp_flat)\n",
    "\n",
    "# Importing inputs, finding complement (non-coding strand), and finding BP regions next to PAM sequences\n",
    "for j, Sequence in enumerate(gene_sequence):\n",
    "    Sequence.upper() # Make sure all basepairs are capitalized for downstream searching\n",
    "    PAM_NT = PAM_NT.upper()\n",
    "    primer_H = primer_H.upper()\n",
    "    gene_name = gene_order[j]\n",
    "    seq_NT = Sequence\n",
    "\n",
    "    BP_Regions = []\n",
    "    PAM_Found = []\n",
    "    PAM_Index = []\n",
    "    # Find the BP regions downstream of PAM sequences\n",
    "    for i, BP in enumerate(seq_NT[0:-20]): # Ignore last 20 BP, as a PAM there doesn't have room for a homology region\n",
    "        if PAM_NT[0] != 'N' and PAM_NT[0] == BP:\n",
    "            if PAM_NT[2] != 'N' and PAM_NT[2] == seq_NT[i+1]:\n",
    "                if PAM_NT[4] != 'N':\n",
    "                    if PAM_NT[4] == seq_NT[i+2]:\n",
    "                        BP_Regions.append(seq_NT[(i+3):(i+23)])\n",
    "                        PAM_Found.append(BP+seq_NT[i+1]+seq_NT[i+2])\n",
    "                        PAM_Index.append(i)\n",
    "                else:\n",
    "                    BP_Regions.append(seq_NT[(i+3):(i+23)])\n",
    "                    PAM_Found.append(BP+seq_NT[i+1]+seq_NT[i+2])\n",
    "                    PAM_Index.append(i)\n",
    "                    \n",
    "    # Extract the sgRNA closest to the n-terminus\n",
    "    output1 = []# Contains BP region design data\n",
    "    output1.append('gene_name' + '\\t' + 'BP_Region' + '\\t' + 'PAM_Found' + '\\t' + 'PAM_Index' + '\\t' + \n",
    "                   'BsaI_Site' + '\\t' + 'primer')\n",
    "   \n",
    "    \n",
    "    for q, item2 in enumerate(BP_Regions):\n",
    "        BP_seq = Seq(str(BP_Regions[q]))\n",
    "        BP_RC = BP_seq.reverse_complement() # Find reverse complement (this is sgRNA sequence)\n",
    "        # Make sure the BsaI site is not in the gRNA\n",
    "        BsaI_seq = Restriction.BsaI.search(BP_seq)\n",
    "        BsaI_RC = Restriction.BsaI.search(BP_RC)\n",
    "        BsaI = str(BsaI_seq) + str(BsaI_RC)\n",
    "        primer = BP_RC + primer_H # This is the BP region plus the homology region to make a complete primer\n",
    "        output = gene_name + '\\t' + BP_RC + '\\t' + PAM_Found[q] + '\\t' + str(PAM_Index[q] + 1) + '\\t' + \\\n",
    "                 BsaI + '\\t' + primer\n",
    "        output1.append(output)\n",
    "        \n",
    "    F = open(sgRNA_path + 'CRISPRi_Design_' + gene_name + '_.txt', 'w') # Write all possible sgRNAs for each gene\n",
    "    # Write data into files and end program.\n",
    "    for item in output1:\n",
    "        F.write(\"%s\\n\" % item)\n",
    "    F.close()\n",
    "\n",
    "Gene_file = glob.glob(sgRNA_path + '*.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Selected sgRNAs for each gene\n",
    "\n",
    "Identifies three potential sgRNAs for each gene of interest by finding all PAM motifs in a gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_parent, sgRNA_parent_blast = pick_sg(Gene_file, Gene_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output for BLAST\n",
    "\n",
    "BLASTing all potential sgRNAs against the E. coli MG1655 genome (details in Methods) to search for potential off-target effects. Ideally, each sgRNA would have one and only one corresponding sequence on the genome. Issues arise when another sequence is (i) adjacent to a valid PAM motif and (ii) >=75% identical to the sgRNA sequence. Even in these cases, mismatches near the PAM can render potential for off-target binding to near-zero, so futher analysis parses these differences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write for BLAST\n",
    "# Initial run had too many sgRNAs for BLASt\n",
    "with open(output_path + '191105_sgRNA_parent_4blast1.txt', 'w') as F:\n",
    "    for item in sgRNA_parent_blast[:200]:\n",
    "        F.write(\"%s\\n\" % item)\n",
    "with open(output_path + '191105_sgRNA_parent_4blast2.txt', 'w') as F:\n",
    "    for item in sgRNA_parent_blast[200:400]:\n",
    "        F.write(\"%s\\n\" % item)\n",
    "with open(output_path + '191105_sgRNA_parent_4blast3.txt', 'w') as F:\n",
    "    for item in sgRNA_parent_blast[400:600]:\n",
    "        F.write(\"%s\\n\" % item)\n",
    "with open(output_path + '191105_sgRNA_parent.txt', 'w') as F:\n",
    "    for item in sgRNA_parent_blast:\n",
    "        F.write(\"%s\\n\" % item)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import BLAST results\n",
    "\n",
    "Following BLASTn analysis, results are read back in and concatenated. Highly similar sequences are checked for the presence of a PAM site and other potential flags for off-target binding found by previous research. Any sgRNAs that fail these tests will be redesigned."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import BLAST data using nBLAST for E. coli MG1655 genome\n",
    "\n",
    "df_blast1 = pd.read_csv(blast_path + 'W3UF9SMM015-191105-Alignment-HitTable1.csv', index_col=0, parse_dates=True)\n",
    "df_blast2 = pd.read_csv(blast_path + 'W3UFRXZE014-191105-Alignment-HitTable2.csv', index_col=0, parse_dates=True)\n",
    "df_blast3 = pd.read_csv(blast_path + 'W3UG1J8F015-191105-Alignment-HitTable3.csv', index_col=0, parse_dates=True)\n",
    "# Concatenate BLAST results\n",
    "df_blast = pd.concat([df_blast1, df_blast2, df_blast3])\n",
    "# Remove all the organisms except MG1655 K12\n",
    "df_filt = df_blast.loc[df_blast['organism'] == 'CP032667.1']\n",
    "# Check how many genes have high homology to more than 1 target\n",
    "df_15 = df_filt.loc[(df_filt['BP_Match'] >= 15)]\n",
    "# Count the number of BLAST hits per sgRNA\n",
    "Blast_count = Counter(df_15.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "Gene_fail, Gene_pass = blast_check(Blast_count, df_15, df_parent)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Redesign sgRNAs where previous sequences were rejected\n",
    "\n",
    "Repeat this process iteratively to narrow down remaining genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_it2, sgRNA_it2_blast = pick_sg_it(Gene_file, Gene_fail)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(output_path + '191105_sgRNA_it2.txt', 'w') as F:\n",
    "    for item in sgRNA_it2_blast:\n",
    "        F.write(\"%s\\n\" % item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import second run of BLAST data\n",
    "df_blast_it2 = pd.read_csv(blast_path + 'W3VA6W0N014-191105-Alignment-HitTable_IT2.csv', \n",
    "                        index_col=0, parse_dates=True)\n",
    "# Remove all the organisms except MG1655 K12\n",
    "df_filt_it2 = df_blast_it2.loc[df_blast_it2['organism'] == 'CP032667.1']\n",
    "# Check how many genes have perfect homology to more than 1 target\n",
    "df_15_it2 = df_filt_it2.loc[(df_filt_it2['BP_Match'] >= 15)]\n",
    "# Count the number of blast hits per sgRNA\n",
    "Blast_count_it2 = Counter(df_15_it2.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "Gene_fail_it2, Gene_pass_it2 = blast_check(Blast_count_it2, df_15_it2, df_it2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Manual BLAST and validation\n",
    "\n",
    "Once the number of remaining genes is small enough, sgRNAs can be designed by finding PAMs and BLASTing adjacent sequences individually."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['fabA_3_210', 'fabA_2_79', 'gyrB_2_67']\n"
     ]
    }
   ],
   "source": [
    "print(Gene_fail_it2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "IT3 = {}\n",
    "IT3['fabA_2_88'] = {}\n",
    "IT3['fabA_3_226'] = {}\n",
    "IT3['gyrB_2_101'] = {}\n",
    "\n",
    "IT3['fabA_2_88']['PAM_index'] = 88\n",
    "IT3['fabA_2_88']['PAM_seq'] = 'CCA'\n",
    "IT3['fabA_2_88']['Sequence'] = 'ATCATCAGCATGTTCGGTGC'\n",
    "\n",
    "IT3['fabA_3_226']['PAM_index'] = 226\n",
    "IT3['fabA_3_226']['PAM_seq'] = 'CCG'\n",
    "IT3['fabA_3_226']['Sequence'] = 'CCCAGGCATCCCGGCATAAC'\n",
    "\n",
    "IT3['gyrB_2_101']['PAM_index'] = 101\n",
    "IT3['gyrB_2_101']['PAM_seq'] = 'CCG'\n",
    "IT3['gyrB_2_101']['Sequence'] = 'GAATACCATGTGGTGCAGAC'\n",
    "\n",
    "df_it3 = pd.DataFrame.from_dict(IT3).T\n",
    "# Took these sequences, manually BLASTed them, and analyzed results below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import BLAST results\n",
    "df_blast_it3 = pd.read_csv(blast_path + 'W3VV4W6F014-191105-Alignment-HitTable_IT3.csv', index_col=0, \n",
    "                           parse_dates=True)\n",
    "# Remove all the organisms except MG1655 K12\n",
    "df_filt_it3 = df_blast_it3.loc[df_blast_it3['organism'] == 'CP032667.1']\n",
    "# Check how many genes have perfect homology to more than 1 target\n",
    "df_15_it3 = df_filt_it3.loc[(df_filt_it3['BP_Match'] >= 15)]\n",
    "# Count the number of BLAST hits per sgRNA\n",
    "Blast_count_it3 = Counter(df_15_it3.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "Gene_fail_it3, Gene_pass_it3 = blast_check(Blast_count_it3, df_15_it3, df_it3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAoAAAADQCAYAAACX3ND9AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3Xm4JHV97/H3x2HfZBsJGRwP6DwqIQHxiAh5DEswAioYwYBGiSGZeK9REKOgyQ14jU8wccOb6HUMhtEgOwTighBE1FwdmBlHtlEhMCEjE2aSsCYIDHzuH1UHDmfOUqdPV1d31+f1PP10V3VX1bdn+tv9PVW/RbaJiIiIiPZ4TtMBRERERERvpQCMiIiIaJkUgBEREREtkwIwIiIiomVSAEZERES0TArAiIiIiJZJARgRERHRMrUWgJLeK+k2SbdKukDSVpL2lLRM0h2SLpK0RZ0xRERERMSz1VYASloAvAcYtb0PMA84AfgY8Cnbi4D7gZPriiEiIiIiNrVZD/a/taQngG2AdcBhwFvK55cCZwGfm24nu+66q0dGRuqLMmIWVqxY8e+25zcdR6eST9FvklMR3VM1n2orAG3/TNLHgXuAR4FrgBXAA7Y3li9bCyyYaV8jIyMsX768rlAjZkXSvzQdw1wkn6LfJKciuqdqPtV5CXgn4BhgT+AXgW2BIyd56aSTEUtaLGm5pOUbNmyoK8yIiIiI1qmzE8ivA3fb3mD7CeBy4CBgR0ljZx73AO6dbGPbS2yP2h6dP39grwxERERE9J06C8B7gAMlbSNJwOHA7cD1wHHla04CrqwxhoiIiIiYoLYC0PYy4FJgJXBLeawlwOnAaZLuBHYBzq0rhoiIiIjYVK29gG2fCZw5YfVdwAF1HjciIiIippaZQCIiIiJapu5xAGNAjZzxtZ4eb83ZR/f0eNEuvf48z0VyIaK6ibmd/KkuZwAjIiIiWiYFYERERETL5BJwRDzLoFwuzaWeiIjO5QxgRERERMukAIyIiIhomRSAERERES2TAjAiIiKiZVIARkRERLRMCsCIASHpvZJuk3SrpAskbSVpT0nLJN0h6SJJWzQdZ0RE9L8UgBEDQNIC4D3AqO19gHnACcDHgE/ZXgTcD5zcXJQRETEoUgBGDI7NgK0lbQZsA6wDDgMuLZ9fChzbUGwRETFAMhB0xACw/TNJHwfuAR4FrgFWAA/Y3li+bC2wYLLtJS0GFgMsXLiw/oCjdoMyYDdk0O6IfpQzgBEDQNJOwDHAnsAvAtsCR07yUk+2ve0ltkdtj86fP7++QCMGhKTnS7pe0uqybe0p5fqzJP1M0qrydlTTsUbUIWcAIwbDrwN3294AIOly4CBgR0mblWcB9wDubTDGiEGyEXif7ZWStgdWSLq2fO5Ttj/eYGwRtcsZwIjBcA9woKRtJAk4HLgduB44rnzNScCVDcUXMVBsr7O9snz8MLCaKZpQRAyjGQtASbtJOlfSN8rlvSWlp2FEhzrJKdvLKDp7rARuocjdJcDpwGmS7gR2Ac6tNfiIISRpBHgZsKxc9YeSbpb0xbL5xWTbLJa0XNLyDRs29CjSiO6pcgbwPOCbFO2OAH4KnFpXQBEtcB4d5JTtM22/xPY+tt9m+zHbd9k+wPaLbB9v+7Ea444YOpK2Ay4DTrX9EPA54IXAfhQ97T8x2XZpVxuDrkoBuKvti4GnAMq2Rk/WGlXEcEtORfQBSZtTFH/n274cwPZ9tp+0/RTwBeCAJmOMqEuVAvC/JO1C2btQ0oHAg7VGFTHcklMRDSvb0p4LrLb9yXHrdx/3sjcCt/Y6toheqNIL+DTgKuCFkv4JmM8zjc4jYvaSUxHNOxh4G3CLpFXlug8BJ0raj+IPtDXAHzQTXkS9ZiwAyy7yvwa8GBDwE9tP1B5ZxJBKTkU0z/b3KPJvoq/3OpaIJlQdB/AAYKR8/f6SsP2l2qKKGH7JqYiIaMyMBaCkL1P0iFrFMw3VDeTHKqIDyamIiGhalTOAo8DetiedYioiZi05FRERjarSC/hW4BfqDiSiRZJTERHRqCpnAHcFbpd0I/D0ILO231BbVBHDLTkVERGNqlIAnlV3EBEtc1bTAURERLtVGQbmBkkvABbZ/kdJ2wDz6g8tYjglpyIiomkztgGU9PsUk9B/vly1APj7OoOKGGbJqYiIaFqVTiDvohgx/SEA23cAz6uyc0k7SrpU0o8lrZb0Kkk7S7pW0h3l/U6dhx8xkDrOqYiIiG6oUgA+ZvvxsQVJm1HOYVrBOcDVtl8C7AusBs4ArrO9CLiuXI5ok7nkVERExJxVKQBvkPQhYGtJRwCXAP8w00aSdgBeTTHZNrYft/0AcAywtHzZUuDYTgKPGGAd5VRERES3VCkAzwA2ALdQTIr9deBPKmy3V7nd30r6oaS/kbQtsJvtdQDl/aSXviQtlrRc0vINGzZUOFzEwOg0pyIiIrqiSi/gp4AvlLfZ7nt/4N22l0k6h1lc7rW9BFgCMDo6mstjMTTmkFMRERFdUWUu4FvYtH3Sg8By4M9s/8cUm64F1tpeVi5fSlEA3idpd9vrJO0OrO8s9IjBNIecioiI6IoqA0F/g2LC+q+UyyeU9w8B5wGvn2wj2/8m6V8lvdj2T4DDgdvL20nA2eX9lR1HHzGYOsqpiIiIbqlSAB5s++Bxy7dI+ifbB0v67Rm2fTdwvqQtgLuAd1C0O7xY0snAPcDxnQQeMcDmklMRERFzVqUA3E7SK8cu5Uo6ANiufG7jdBvaXgWMTvLU4bOKMmK4dJxTETH4Rs742rOW15x9dEORRJtVKQB/D/iipLEfqIeBk8sevX9eW2QRwys5FRERjarSC/gm4JclPRdQOZbfmItriyxiSCWnIpon6fnAl4BfAJ4Cltg+R9LOwEXACLAGeLPt+5uKM6IuVcYBBMD2gxN+qCJiDpJTEY3aCLzP9kuBA4F3SdqbzFYVLVG5AIyIiBgWttfZXlk+fphiqtIFZLaqaIlpC0BJz5F0UK+CiRh2yamI/iNpBHgZsIzMVhUtMW0BWM5Y8IkexRIx9OaSU5J2lHSppB9LWi3pVZJ2lnStpDvK+526HHLEUCs7Y10GnGr7oarb2V5ie9T26Pz58+sLMKImVS4BXyPpTZJUezQR7dBpTp0DXG37JcC+FJes0l4pokOSNqco/s63fXm5+r5ylioyW1UMsyrDwJwGbAs8KelRQIBt71BrZBHDa9Y5JWkH4NXA71C8+HHgcUnHAIeUL1sKfBs4va7AI4ZF+QfYucBq258c99RVZLaqaIEqw8Bs34tAItqiw5zaC9gA/K2kfYEVwClMaK8kacr2SsBigIULF3YUd8SQORh4G8VMPKvKdR+iKPwyW1UMvRkLwPKvpLcCe9r+SDl20u62b6w9uogh1GFObQbsD7zb9jJJ5zCLy722lwBLAEZHR9159BHDwfb3KM6+TyazVcXQq9IG8LPAq4C3lMuPAH9dW0QRw6+TnFoLrB2bPg64lKIgTHuliIiYtSoF4Cttvwv4OUA5IvoWtUYVMdxmnVO2/w34V0kvLlcdDtzOM+2VIO2VIiKioiqdQJ6QNA8wgKT5FNPmRERnOs2pdwPnS9oCuAt4B8UfcWmvFBERs1KlAPwMcAWwm6SPAscBf1JrVBHDraOcsr0KGJ3kqbRXioiIWanSC/h8SSt45kfmWNur6w0rYnglpyIiomlVzgACbAOMXbLaur5wIlojORUREY2ZsROIpD+lGGB2Z2BXinHIcgk4okPJqYiIaFqVM4AnAi+z/XMASWcDK4E/qzOwiCGWnIqIiEZVGQZmDbDVuOUtgX+uJZqIdlhDciqi6yTtJOlXmo4jYhBUOQP4GHCbpGsp2isdAXxP0mcAbL+nxvgihlFyKqJLJH0beAPF79kqYIOkG2yf1mhgEX2uSgF4RXkb8+16QolojeRURPc81/ZDkn4P+FvbZ0q6uemgZmPkjK89a3nN2Uc3FMngm/hvCfn3nEqVYWCW9iKQiLZITkV01WblNIhvBv646WAiBkWVNoARERH96sPAN4E7bd8kaS/gjoZjiuh7VccBjIiI6EfrbD/d8cP2XZI+2WRAEYNg1mcAJW0lKfONRnRJcipiTv5PxXURMU6lM4DlxPWvoRi/7DeA7wKX1BhXxFBLTkXMjaRXAQcB8yWN7/G7A8UsOxExjWkLQEmvBt4CHA3cCBwM7Gn7v3sQW8TQSU5FdM0WwHYUv2Pbj1v/EHBcIxFFDJApC0BJa4F7gM8B77f9sKS780MV0ZnkVET32L4BuEHSebb/pel4IgbNdG0ALwMWAL8FvF7SthSD1kZEZ5JTEd23paQlkq6R9K2x20wbSfqipPWSbh237ixJP5O0qrwdVW/oEc2ZsgC0fQowAnwSOBT4KUVbizdL2q434UUMj+RURC0uAX4I/Anw/nG3mZwHvHaS9Z+yvV95+3rXoozoM9O2AbRt4FvAtyRtDhwJnAB8Fti1ygHKxu7LgZ/Zfp2kPYELgZ2BlcDbbD/e+VuIGBzdyKmIeJaNtj83241sf0fSSPfDiRgMlccBtP0EcBVwlaStZ3GMU4DVFD2zAD5G8RfWhZL+L3AyRZuoiFaZQ05FxDP+QdL/pJhe8bGxlbb/s8P9/aGkt1OcuHif7fsne5GkxcBigIULF3Z4qMllOrPohek6gVzP1O2TDBw+084l7UHR2/GjwGmSBBxG0QsSYClwFikAowW6kVMRsYmTyvvxl30N7NXBvj4HfKTc/iPAJ4DfneyFtpcASwBGR0fTljcGznRnAP9oknUHAh8A1lfc/6fL14910d8FeMD2xnJ5LUWj+Ig26EZORcQ4tvfs4r7uG3ss6QvAV7u174h+M2UBaHvF2GNJvwb8L2BL4J22vzHTjiW9Dlhve4WkQ8ZWT3aoKbav7fR6RBPmmlMRsanycu0mbH+pg33tbntdufhG4NbpXh8xyGYaCPo3KH6kfg581Pb1s9j3wcAbym70W1G0Afw0sKOkzcqzgHsA9062cU6vxzCaY05FxKZeMe7xVhRNKVYC0xaAki4ADgF2LcfoPBM4RNJ+FCcm1gB/UEO8EX1hujaANwHzgb8Evl+u23/sedsrp9ux7Q8CHyy3OwT4I9tvlXQJxSjtF1K03bhybm8hYjDMNaciYlO23z1+WdJzgS9X2O7ESVaf2624IvrddGcA/wt4hKJYexPPvnxris4cnTgduFDSn1GM3ZSEi7aoK6ci4hn/DSxqOoi6VekpPPE13epJXNd+o7emawN4SLcOYvvbwLfLx3cBB3Rr3xGDops5FREFSf/AM23J5wEvBS5uLqKIwVB5HMAxko4APmD7iBriiWid5FTEnHx83OONwL/YXttUMBGDYsqp4CQdJumnkh6R9HeS9pa0HDibjNsXMWvdyClJ8yT9UNJXy+U9JS2TdIekiyRtUed7iOg3tm8Afkwx3NhOQGaWiqhgygKQYgDMxRRj910K/AD4su2X2768F8FFDJlu5NTYzDpjxmbWWQTcTzGzTkRrSHozcCNwPPBmYJmk45qNKqL/TXcJ2GXbPYC/l7TB9jk9iCliWM0ppzKzTsSk/hh4he31AJLmA/9I8UdWq0zWMSRiKtMVgDtK+s1xyxq/nLOAEbM215zqeGadDKweQ+w5Y8Vf6T+Y/upWRDB9AXgD8Poplg2kAIyYnY5zaq4z62Rg9RhiV0v6JnBBufxbwNcbjCdiIEw3DMw7ehlIxLCbY07NaWadiGEj6UXAbrbfX55J/1WKP4q+D5zfaHARA2DGYWAknTbJ6geBFbZXdT+kiOHWSU5lZp2ITXwa+BA83XzicgBJo+Vzr59604io0k5iFHgnRduiBRTtiA4BviDpA/WFFjG0uplTp1N0CLmTok1gZtaJthixffPElbaXAyO9DydisFQZCHoXYH/bjwBIOpOid9WrgRXAX9QXXsRQmlNOZWadCKBoCjGVrXsWRY+kh290W5UzgAt59sCaTwAvsP0o8FgtUUUMt+RUxNzdJOn3J66UdDLFH1IRMY0qZwC/AvxA0ljbojcAF0jaFri9tsgihldyKmLuTgWukPRWnin4RoEtgDc2FlXEgJixALT9EUlf55keVn9QtrEAeGudwUUMo+RUxNzZvg84SNKhwD7l6q/Z/laDYUUMjCq9gF8I3FqOP3YoRcLdafuB+sOLGD7JqYjusX09cH3TcUQMmiptAC8DnizHXPoCsCfFJayI6ExyKiIiGlWlAHyqHGT2N4FzbL8X2L3esCKGWnIqIiIaVaUAfELSicDbga+W6zavL6SIoZecimiYpC9KWi/p1nHrdpZ0raQ7yvudmowxok5VCsB3AK8CPmr7bkl7An9Xb1gRQy05FdG884DXTlh3BnCd7UXAdeVyxFCq0gv4duA945bvBs6uM6iIYZacimie7e9IGpmw+hiKWXkAllIMuH56z4KK6KEq4wBGRES0wW621wHYXifpeVO9UNJiimkcWbhwYY/C676JM4ysOfvohiKpTxveYyeqXAKOiIiIcWwvsT1qe3T+/PlNhxMxazMWgJKOr7IuIqpJTkX0rfsk7Q5Q3q9vOJ6I2lQ5A/jBiusioprkVER/ugo4qXx8EnDlNK+NGGhTtgGUdCRwFLBA0mfGPbUDsLHuwCKGTXIqon9IuoCiw8euktYCZ1J0xrpY0snAPUDOzMfQmq4TyL3AcoqJ6leMW/8w8N46g4oYUsmpiD5h+8Qpnjq8p4FENGTKAtD2j4AfSfqK7Sd6GFPEUEpORUSTJvaGjXarMgzMAZLOAl5Qvl6Abe9VZ2ARQyw5FRERjapSAJ5LcXlqBfBkveFEtEJyKiIiGlWlAHzQ9jdqjySiPZJTERHRqCoF4PWS/hK4HHhsbKXtlbVFFTHcklMREdGoKgXgK8v70XHrDBzW/XAiWiE5FRHRkMk6w7RxergZC0Dbh3ayY0nPB74E/ALwFLDE9jmSdgYuAkaANcCbbd/fyTEiBlGnORUREdEtVaaC203SuZK+US7vXQ6SOZONwPtsvxQ4EHiXpL2BM4DrbC8CriuXI1pjDjkVERHRFVWmgjsP+Cbwi+XyT4FTZ9rI9rqxNk22HwZWAwuAY4Cl5cuWAsfOLuSIgXceHeRUREREt1QpAHe1fTHFZVxsb2SWQ1dIGgFeBiwDdrO9rtzXOuB5U2yzWNJyScs3bNgwm8NF9Ls551RERMRcVCkA/0vSLhSN1JF0IPBg1QNI2g64DDjV9kNVt7O9xPao7dH58+dX3SxiEMwppyIiIuaqSi/g04CrgBdK+idgPnBclZ1L2pyi+Dvf9uXl6vsk7W57naTdgfUdxB0xyGadU+lUFRGDZGJP2zb2su130xaAkp4DbAX8GvBiiimrflJlHlNJopjxYLXtT4576irgJODs8v7KzkKPGDxzyKmxTlUrJW0PrJB0LfA7FJ2qzpZ0BkWnqtNrewMRETEUpr0EbPsp4BO2N9q+zfats5jE/mDgbcBhklaVt6MoCr8jJN0BHFEuR7RCpzmVTlUREdFNVS4BXyPpTcDltl11x7a/R3F2YzKHV91PxBDqKKfGTNepStKUnaqAxQALFy7sMOyIiBgWVdsAbgtslPRziqLOtneoNbKI4dVxTk3sVFW0tJiZ7SXAEoDR0dFZF50RETFcqswEsn0vAoloi05zKp2qIiLq0cZOK1XOACJpJ2ARReN1AGx/p66gIobdbHMqnaoiIqKbZiwAJf0ecAqwB7CKYlq375OJ6yM60mFOjXWqukXSqnLdhygKv4vLqeTuAY6vK+6ItpC0BniYYoD2jbZHm40oovuqnAE8BXgF8APbh0p6CfDhesOKGGqzzql0qorouUNt/3vTQUTUpcpMID+3/XMASVva/jHF+GUR0ZnkVERENKrKGcC1knYE/h64VtL9wL31hhUx1JJTEf3NFMM1Gfh82Ys+YqhU6QX8xvLhWZKuB54LXF1rVBFDLDkV0fcOtn1vOa7mtZJ+PLGTVtWxNSf2Lu13gxZvXdrQK3jKAlDSVsA7gRcBtwDn2r6hV4FFDJvkVMRgsH1veb9e0hXAAcB3JrwmY2vGQJuuDeBSYJTih+pI4BM9iShieCWnIvqcpG3L+baRtC3wGuDWZqOK6L7pLgHvbfuXASSdC9zYm5AihlZyKqL/7QZcUc6ysxnwFdtpohFDZ7oC8OkJ6m1vrDrlVERMKTkV0eds3wXs23QcEXWbrgDcV9JD5WMBW5fLmQs4ojPJqYgYOm3tOFLlffdz55EpC0Db83oZSMSwS05FRES/qDIQdEREREQMkRSAERERES2TAjAiIiKiZVIARkRERLRMCsCIiIiIlkkBGBEREdEy040DGH2irWMsRURERD1yBjAiIiKiZVIARkRERLRMCsCIiIiIlkkBGBEREdEyKQAjIiIiWiYFYERERETLpACMiIiIaJmMA9iBjMsXERHRHt363Z+4nzVnH93RsSbbbrZyBjAiIiKiZVIARkRERLRMIwWgpNdK+omkOyWd0UQMEcMkORXRPcmnaIOetwGUNA/4a+AIYC1wk6SrbN/e61iifzTRrrIbbSj6QXIqonuST9EWTZwBPAC40/Zdth8HLgSOaSCOiGGRnIronuRTtEITvYAXAP86bnkt8MqJL5K0GFhcLj4i6SddjmNX4N+7vM9+O2Yb3mPHx9THOj7eizvesh4z5lQP8mkmXf9czOH/r1tq+awP4/uq8J5e0M3jzVG3f6Oa+E6ci0H7XPf1v+8k7/tZ8Xb67zLDdpXyqYkCUJOs8yYr7CXAktqCkJbbHq1r//1wzDa8xyaOKWl5r45V0Yw5VXc+zaSJz0XdhvE9wfC+r1no6m/UoP17Jt569VO8TVwCXgs8f9zyHsC9DcQRMSySUxHdk3yKVmiiALwJWCRpT0lbACcAVzUQR8SwSE5FdE/yKVqh55eAbW+U9IfAN4F5wBdt39brOGjmclivj9mG99jEMRu7lDqZPsqp6fTVv1mXDON7guF9X5XUkE+D9u+ZeOvVN/HK3qRpQ0REREQMscwEEhEREdEyKQAjIiIiWqaVBWCvp/mR9EVJ6yXdWvexyuM9X9L1klZLuk3SKT045laSbpT0o/KYH677mOVx50n6oaSv9uh4ayTdImlVHw4H03d6/dnvhaY+63Vq4jujDfpxSrnJclLSzpKulXRHeb9TuV6SPlPGf7Ok/Xsc66Sfyz6Od9LvhrJD0bIy3ovKzkVI2rJcvrN8fqSX8bauABw3zc+RwN7AiZL2rvmw5wGvrfkY420E3mf7pcCBwLt68B4fAw6zvS+wH/BaSQfWfEyAU4DVPTjOeIfa3q9fxnLqc+fR289+LzT1Wa9TE98ZQ62h35oqzmPTnDwDuM72IuC6chmK2BeVt8XA53oU45ipPpf9Gu9U3w0fAz5Vxns/cHL5+pOB+22/CPhU+bqeaV0BSAPT/Nj+DvCfdR5jwvHW2V5ZPn6YokBaUPMxbfuRcnHz8lZrDyNJewBHA39T53Gic73+7PdCE5/1ujXxndECfTml3BQ5eQywtHy8FDh23PovlZ/5HwA7Stq9N5FO+7ns13in+m44DLh0injH3selwOGSJhuIvBZtLAAnm+ZnaL/oylPKLwOW9eBY8yStAtYD19qu+5ifBj4APFXzccYzcI2kFSqmgooWauCz3jO9/M4YcoP0W7Ob7XVQFF3A88r1ffMeJnwu+zbeid8NwD8DD9jeOElMT8dbPv8gsEuvYm1jAVhpmp9hIGk74DLgVNsP1X0820/a3o9i5PwDJO1T17EkvQ5Yb3tFXceYwsG296e41PAuSa/u8fGjD/Tys95Lvf7OGHLD8FvTF+9hFp/LxuOd+N0AvHSamBqNt40FYCum+ZG0OUXCnG/78l4e2/YDwLept+3XwcAbJK2huLRymKS/q/F4ANi+t7xfD1xBkeDRUj36rPdEk98ZQ2qQfmvuG7tUWt6vL9c3/h6m+Fz2bbxjxn03HEhxKXps4o3xMT0db/n8c+lhk5k2FoBDP81P2YbgXGC17U/26JjzJe1YPt4a+HXgx3Udz/YHbe9he4Ti//Bbtn+7ruMBSNpW0vZjj4HXAEPTuzWq6fVnvRea+M5ogUH6rbkKOKl8fBJw5bj1by971x4IPDh26bUXpvlc9mu8k303rAauB46bIt6x93Ecxe9Y785Y2m7dDTgK+CnFtfk/7sHxLgDWAU9QVPwn13y8X6U4jXwzsKq8HVXzMX8F+GF5zFuBP+3h/+chwFd7cJy9gB+Vt9t68dkZ9FuvP/s9ek+NfdZrfE89/85ow63XvzUVY9okJynanV0H3FHe71y+VhQ9mf8ZuAUY7XGsk34u+zjeSb8byt+OG4E7gUuALcv1W5XLd5bP79XLeDMVXERERETLtPEScERERESrpQCMiIiIaJkUgBEREREtkwIwIiIiomVSAEZERES0TArALpP0pKRVkm6VdImkbcY990ZJlvSScetGynUfGbduV0lPSPqrSfb/O5Ot7yDOr4+NVxTRz5JTEd2TfIoxKQC771Hb+9neB3gceOe4504EvkcxIOh4dwGvG7d8PMU4c7WxfZSLkcoj+l1yKqJ7kk8BpACs23eBF8HTcxkeTDHo5sTkehRYLWm0XP4t4OKZdi7pBZKuk3Rzeb+wXL+bpCsk/ai8HTTJtmvKv+JGJK2W9AVJt0m6phzBfOLrXyjpB5JukvS/JT0y7rn3l+tvlvThct2U+y33dbWkFZK+O/6vzYgZJKeSU9E9yacW51MKwJqomNfvSIrRyAGOBa62/VPgPyXtP2GTC4ETJO0BPEm1+Qv/CviS7V8Bzgc+U67/DHCD7X2B/Zn5L7VFwF/b/iXgAeBNk7zmHOAc268YH5uk15TbHwDsB7xc0qtn2O8S4N22Xw78EfDZCu81Wi45Ne1+k1MxK8mnaffbinzabOaXxCxtLWlV+fi7FPMYQnFq/dPl4wvL5ZXjtrsa+AhwH3BRxWO9CvjN8vGXgb8oHx8GvB3A9pPAgzPs527bYzGvAEamONax5eOvAB8vH7+mvP2wXN6OIqnumWy/5V+ZBwGXSBrb95YzxBftlpxKTkX3JJ+ST0AKwDo8anu/8Ssk7ULxgd9HkoF5gCUT3qmSAAABfUlEQVR9YOw1th+XtAJ4H/BLwOs7OHan8/o9Nu7xk8Amp9enIeDPbX/+WSulkSn2+xzggYn/RhHTSE6RnIquST6RfIJcAu6V4yhOg7/A9ojt5wN3U0x0Pd4ngNNt/0fF/f4/nmmr8VaKxrtQTI79PwAkzZO0w5yiL/yAZ06Pj28f8k3gd8u/mpC0QNLzptqJ7YeAuyUdX75ekvbtQnzRLsmpUnIquiD5VGpTPqUA7I0TgSsmrLsMeMv4FbZvs710Fvt9D/AOSTcDbwNOKdefAhwq6RaK09q/1FHUz3YqcJqkG4HdKU/Z276G4nT798vjXQpsP8O+3gqcLOlHFG0/julCfNEuyalnS07FXCSfnq0V+SS70zOy0SYqxop61LYlnQCcaHsokyKiF5JTEd2TfJq9tAGMql4O/JWKVrEPAL/bcDwRgy45FdE9yadZyhnAiIiIiJZJG8CIiIiIlkkBGBEREdEyKQAjIiIiWiYFYERERETLpACMiIiIaJn/D48gOhVC8zHEAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 648x216 with 3 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "Filter_list = Gene_fail + Gene_fail_it2 + Gene_fail_it3\n",
    "# Combine together parent sgRNAs\n",
    "df_parent_filter = pd.concat([df_parent, df_it2, df_it3])\n",
    "# Remove sgRNAs that failed previous iterations\n",
    "df_parent_filter = df_parent_filter.drop(Filter_list)\n",
    "# Remove any entirely redundant entries (redesigning one sgRNA can duplicate others)\n",
    "df_parent_filter = df_parent_filter.loc[~df_parent_filter.index.duplicated(keep='first')]\n",
    "\n",
    "# Filtering can still allow for genes to have more than 3 guides in some cases\n",
    "# Count which genes have more than three guides\n",
    "sg_gene = {}\n",
    "sg_loc = {}\n",
    "for ID in df_parent_filter.index:\n",
    "    gene, rep, pam = ID.split('_')\n",
    "    if gene not in sg_gene.keys():\n",
    "        sg_gene[gene] = 0\n",
    "        sg_gene[gene] += 1\n",
    "    else:\n",
    "        sg_gene[gene] += 1\n",
    "    if rep not in sg_loc.keys():\n",
    "        sg_loc[rep] = 0\n",
    "        sg_loc[rep] += 1\n",
    "    else:\n",
    "        sg_loc[rep] += 1\n",
    "        \n",
    "plt.figure(figsize=(9, 3))\n",
    "plt.subplot(131)\n",
    "bins = np.arange(0, 6)\n",
    "plt.hist(sg_gene.values(), bins=bins)\n",
    "plt.xticks(bins+.5, bins)\n",
    "plt.xlabel('PAM loc in gene')\n",
    "plt.ylabel('Parent sgRNA per gene')\n",
    "\n",
    "plt.subplot(132)\n",
    "index = np.arange(len(sg_loc.values()))\n",
    "plt.bar(index, sg_loc.values())\n",
    "plt.xticks(index, sg_loc.keys())\n",
    "plt.xlabel('PAM loc in gene')\n",
    "plt.ylabel('Parent sgRNA per gene')\n",
    "\n",
    "plt.subplot(133)\n",
    "bins = np.arange(0, 350, 10)\n",
    "plt.hist(list(df_parent_filter['PAM_index']), bins=bins)\n",
    "plt.xlabel('PAM loc in gene')\n",
    "plt.ylabel('Counts')\n",
    "\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nadE has 1 sgRNAs\n",
      "ribF has 2 sgRNAs\n",
      "rpsE has 2 sgRNAs\n",
      "tyrA has 2 sgRNAs\n",
      "fmt has more than 3 sgRNAs\n",
      "pykF has 1 sgRNAs\n",
      "pykA has 2 sgRNAs\n",
      "infA has 2 sgRNAs\n",
      "purU has 2 sgRNAs\n",
      "gpmA has 2 sgRNAs\n",
      "pyrH has 1 sgRNAs\n"
     ]
    }
   ],
   "source": [
    "# One gene has more than 3 sgRNAs\n",
    "for key in sg_gene.keys():\n",
    "    if sg_gene[key] > 3:\n",
    "        print(key + ' has more than 3 sgRNAs')\n",
    "    elif sg_gene[key] == 1:\n",
    "        print(key + ' has 1 sgRNAs')\n",
    "    elif sg_gene[key] == 2:\n",
    "        print(key + ' has 2 sgRNAs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PAM_index                     104\n",
      "PAM_seq                       CCC\n",
      "Sequence     TCCTGCCGGTCGGTCTGGCT\n",
      "Name: fmt_2_104, dtype: object\n",
      "\n",
      "PAM_index                     222\n",
      "PAM_seq                       CCA\n",
      "Sequence     GCAGTTCGGCGACCAGTTGC\n",
      "Name: fmt_3_222, dtype: object\n",
      "\n",
      "PAM_index                     104\n",
      "PAM_seq                       CCC\n",
      "Sequence     TCCTGCCGGTCGGTCTGGCT\n",
      "Name: fmt_1_104, dtype: object\n",
      "\n",
      "PAM_index                     145\n",
      "PAM_seq                       CCC\n",
      "Sequence     GCCAGAACTTTAACCGGGCT\n",
      "Name: fmt_2_145, dtype: object\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Investigate fmt\n",
    "for ID in df_parent_filter.index:\n",
    "    gene = ID.split('_')[0]\n",
    "    if gene == 'fmt':\n",
    "        print(df_parent_filter.loc[ID])\n",
    "        print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fmt_1_104 and fmt_2_104 are redundant\n",
    "df_parent_final = df_parent_filter.drop('fmt_2_104')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Making sgRNA mismatch libraries\n",
    "\n",
    "Titrated CRISPRi relies on using partially mismatched sgRNAs stemming from a single on-target parent. This code redesigns each parent sgRNA using mismatch strategies described in the publication."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Permute sgRNAs for titrated knockdown\n",
    "sgRNAs = []\n",
    "sgRNA_name = []\n",
    "for ID in df_parent_final.index:\n",
    "    sgRNAs.append(df_parent_final['Sequence'].loc[ID])\n",
    "    sgRNA_name.append(ID)\n",
    "\n",
    "permutes = {}\n",
    "for i, sg in enumerate(sgRNAs):\n",
    "    permutes[sgRNA_name[i]] = permute_sgRNA(i, sg, sgRNA_name[i], 1, 13, Promoter, LoopTerm, \n",
    "                                            mut_pos1, mut_pairs, output_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compile all complete sgRNA sequences\n",
    "\n",
    "Homology region as well as flanking sequences must be included to properly synthesize an sgRNA library. This compiles all previously designed sequences, various negative control guides (both empty and randomly generated), and cell death control guides (initially targeted purE, but future analysis identified other cell death controls that were more effective)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add all \"normal\" sequences to fasta order\n",
    "fasta_order = []\n",
    "for gene in sorted(permutes.keys()):\n",
    "    for MM in permutes[gene]:\n",
    "        fasta_order.append('>'+gene + '_' + MM)\n",
    "        fasta_order.append(permutes[gene][MM])\n",
    "        \n",
    "# Manually add a few more sequences\n",
    "# negC none control 5 times\n",
    "fasta_order.append('>negC_none_1')\n",
    "fasta_order.append(Promoter + LoopTerm)\n",
    "fasta_order.append('>negC_none_2')\n",
    "fasta_order.append(Promoter + LoopTerm)\n",
    "fasta_order.append('>negC_none_3')\n",
    "fasta_order.append(Promoter + LoopTerm)\n",
    "fasta_order.append('>negC_none_4')\n",
    "fasta_order.append(Promoter + LoopTerm)\n",
    "fasta_order.append('>negC_none_5')\n",
    "fasta_order.append(Promoter + LoopTerm)\n",
    "fasta_order.append('>negC_none_6')\n",
    "fasta_order.append(Promoter + LoopTerm)\n",
    "fasta_order.append('>negC_none_7')\n",
    "fasta_order.append(Promoter + LoopTerm)\n",
    "fasta_order.append('>negC_none_8')\n",
    "fasta_order.append(Promoter + LoopTerm)\n",
    "fasta_order.append('>negC_none_9')\n",
    "fasta_order.append(Promoter + LoopTerm)\n",
    "fasta_order.append('>negC_none_10')\n",
    "fasta_order.append(Promoter + LoopTerm)\n",
    "\n",
    "# purE dead control\n",
    "fasta_order.append('>purE_1_9_1')\n",
    "fasta_order.append(Promoter + 'CGACACGCGCCGGATTATTG' + LoopTerm)\n",
    "fasta_order.append('>purE_1_9_2')\n",
    "fasta_order.append(Promoter + 'CGACACGCGCCGGATTATTG' + LoopTerm)\n",
    "fasta_order.append('>purE_1_9_3')\n",
    "fasta_order.append(Promoter + 'CGACACGCGCCGGATTATTG' + LoopTerm)\n",
    "fasta_order.append('>purE_1_9_4')\n",
    "fasta_order.append(Promoter + 'CGACACGCGCCGGATTATTG' + LoopTerm)\n",
    "fasta_order.append('>purE_1_9_5')\n",
    "fasta_order.append(Promoter + 'CGACACGCGCCGGATTATTG' + LoopTerm)\n",
    "\n",
    "fasta_order.append('>purE_2_19_1')\n",
    "fasta_order.append(Promoter + 'ATCACGATGGCGACACGCGC' + LoopTerm)\n",
    "fasta_order.append('>purE_2_19_2')\n",
    "fasta_order.append(Promoter + 'ATCACGATGGCGACACGCGC' + LoopTerm)\n",
    "fasta_order.append('>purE_2_19_3')\n",
    "fasta_order.append(Promoter + 'ATCACGATGGCGACACGCGC' + LoopTerm)\n",
    "fasta_order.append('>purE_2_19_4')\n",
    "fasta_order.append(Promoter + 'ATCACGATGGCGACACGCGC' + LoopTerm)\n",
    "fasta_order.append('>purE_2_19_5')\n",
    "fasta_order.append(Promoter + 'ATCACGATGGCGACACGCGC' + LoopTerm)\n",
    "\n",
    "fasta_order.append('>purE_3_47_1')\n",
    "fasta_order.append(Promoter + 'CATGGTAGCCCAGTCGCTTT' + LoopTerm)\n",
    "fasta_order.append('>purE_3_47_2')\n",
    "fasta_order.append(Promoter + 'CATGGTAGCCCAGTCGCTTT' + LoopTerm)\n",
    "fasta_order.append('>purE_3_47_3')\n",
    "fasta_order.append(Promoter + 'CATGGTAGCCCAGTCGCTTT' + LoopTerm)\n",
    "fasta_order.append('>purE_3_47_4')\n",
    "fasta_order.append(Promoter + 'CATGGTAGCCCAGTCGCTTT' + LoopTerm)\n",
    "fasta_order.append('>purE_3_47_5')\n",
    "fasta_order.append(Promoter + 'CATGGTAGCCCAGTCGCTTT' + LoopTerm)\n",
    "\n",
    "# Generate random sequences for negC\n",
    "for q in np.arange(45):\n",
    "    sg_rand = np.random.choice(['A', 'T', 'C', 'G'], 20)\n",
    "    sg_rand = ''.join(sg_rand)\n",
    "    fasta_order.append('>negC_rand_' + str(q))\n",
    "    fasta_order.append(Promoter + sg_rand + LoopTerm)\n",
    "\n",
    "with open(output_path + '20191111_sgRNA_lib.fasta', 'w') as f:\n",
    "    for item in fasta_order:\n",
    "        f.write(\"%s\\n\" % item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "negC\n",
      "purE\n"
     ]
    }
   ],
   "source": [
    "# Sanity check\n",
    "gene_complete = []\n",
    "for row in fasta_order:\n",
    "    if row[0] == '>':\n",
    "        gene1 = row.split('_')[0][1:]\n",
    "        gene_complete.append(gene1)\n",
    "        \n",
    "gene_complete = set(gene_complete)\n",
    "\n",
    "for g1 in gene_complete:\n",
    "    if g1 in df_genes.index:\n",
    "        continue\n",
    "    else:\n",
    "        print(g1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
